Added wireplumber mixer widget

AKArien opened the pull request:
The widget gives an interface for controlling the volume and muting of sinks, sources and output streams, as well as selection of the default sinks and sources (input/output devices).
By default, upon changes to a volume / mute, it pops up an interactible bubble where volume/mute can be changed again.
Has scroll to change volume and possibility of showing the controls of only a single track at a time, through the « face/quick action target »
Has the following settings :
- Popup on change : enable or disable the « popup on change » behavior. If on, the bubble described above will appear when the « face » volume or mute changes.
- Timeout for hiding popups
- Length of sliders
- Scroll sensitivity (defaults to 1), and inversion
- The buttons to mute and select default are on the right by default, but can be switched to the left
- Selection of what should be the « quick target » for muting and to be shown for a quick control with the corresponding actions : last changed volume, default sink, default source.
- Assignment of view of the full mixer, quick target and muting between left, middle and right click (mute not available on left click)
- icon size

It’s build is behind a new build option and uses the wireplumber C library.
Code for icon selection and the custom gtk scale have been moved out of the pulseaudio volume widget for re-use.
This PR slightly alters the volume widget, by changing the « volume out of range » icon the « emblem-unreadable » icon instead of « audio-volume-muted »
Additionally, the compile-time warning that was emitted if libpulse was not found and the volume widget not built has been moved to a meson notice.

ammen99 commented on src/panel/panel.cpp:
```diff
@@ -164,6 +167,17 @@ class WayfirePanel::impl
+#ifdef HAVE_WIREPLUMBER
+            return Widget(new WayfireWireplumber());
+#else
+    #warning "Wireplumber not found, mixer widget will not be available."
```
I think this warning, if emitted at all, should be emitted from meson.build, not from the source file.

ammen99 commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,1006 @@
+    // sets up the « widget part »
+
+    button = std::make_unique<WayfireMenuButton>("panel");
+    button->get_style_context()->add_class("volume");
```
I think we should use a different class for wireplumber.

AKArien replied:
Oh hey, i just changed that :P

ammen99 commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,1006 @@
+
+    /* Setup popover */
+    popover->set_child(master_box);
+    popover->get_style_context()->add_class("volume-popover");
```
same here, should be a different name for proper css styling support.

AKArien commented:
Should be good !'m not running intitulé any bugs in my own use.
The one thing i’m not quite happy about is the hacky way of « keeping the popover out but swapping it’s content » on click by closing it right before, so it gets opened again by the click event (line 469/490).
It might also help with readability to split it into multiple files ?

NamorNiradnug commented on src/panel/widgets/wireplumber.hpp:
```diff
@@ -0,0 +1,152 @@
+
+static std::vector<WayfireWireplumber*> widgets;
+
+static void init_wp();
```
Do not use `static` in headers. This function should go into `wireplumber.cpp`

NamorNiradnug commented on src/util/meson.build:
```diff
@@ -6,6 +6,7 @@ util = static_library(
         'wf-autohide-window.cpp',
         'wf-popover.cpp',
         'css-config.cpp',
+		'animated-scale.cpp'
```
Use spaces here

NamorNiradnug commented on meson.build:
```diff
@@ -33,6 +34,10 @@ if libpulse.found()
 endif
 
+if wireplumber.found()
+	add_project_arguments('-DHAVE_WIREPLUMBER=1', language: 'cpp')
```
Use spaces here

NamorNiradnug commented on src/panel/widgets/wireplumber.hpp:
```diff
@@ -0,0 +1,152 @@
+  public:
+    void init(Gtk::Box *container) override;
+
+    WfOption<double> scroll_sensitivity{"panel/wp_scroll_sensitivity"};
```
The fields should be private.

NamorNiradnug commented on src/panel/widgets/wireplumber.hpp:
```diff
@@ -0,0 +1,152 @@
+    Gtk::Popover *popover;
+
+    /*
+     *   the « face » is the representation of the audio channel that shows it’s volume level on the widget
```
Please don't use unicode characters in code and comments. `« face »`→ `"face"`

NamorNiradnug commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,984 @@
+            }
+
+            // if the control is not for a sink or source (non WfWpControlDevice), don’t try to set status
+            const gchar *type = wp_pipewire_object_get_property(obj, PW_KEY_MEDIA_CLASS);
```
Make this variable of type `std::string_view`. Then you can compare it with `==` without all these `g_strcmp0`s.

NamorNiradnug commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,984 @@
+        if (g_strcmp0(type, "Audio/Sink") == 0)
+        {
+            which_box = &(widget->sinks_box);
+            control   = new WfWpControlDevice(obj, widget);
```
Don't manage memory with `new/delete`.

Probably `objects_to_control` should own the objects through `std::unique_ptr`.

NamorNiradnug commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,984 @@
+{
+    // adds a new widget to the appropriate section
+
+    WpPipewireObject *obj = (WpPipewireObject*)object;
```
`auto *`

NamorNiradnug commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,984 @@
+
+    WpPipewireObject *obj = (WpPipewireObject*)object;
+
+    const gchar *type = wp_pipewire_object_get_property(obj, PW_KEY_MEDIA_CLASS);
```
`std::string_view`

NamorNiradnug commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,984 @@
+    reload_config();
+
+    // boxes hierarchy and labeling
+    auto r1 = Gtk::Orientation::HORIZONTAL;
```
Please don't use such aliases. It decreases code readability.

ammen99 commented:
> It might also help with readability to split it into multiple files ?

Feel free to do so if you find it better. We already do that for window-list, simply create a subdirectory and place all files in it.

ammen99 commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,983 @@
+     * We re-use the core, manager and all other objects
+     */
+
+    WpCommon::widgets.push_back(this);
```
Once you have a singleton class for the WpCommon stuff, you can move all this code in a method there (`add_widget` / `remove_widget`)

ammen99 commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,983 @@
+void WayfireWireplumber::reload_config()
+{
+    // big matching operation
+    WfOption<std::string> str_face_choice{"panel/wp_face_choice"};
```
These can be marked as static to avoid reloading them every time. Their values will change automatically when the config changes.

AKArien commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,983 @@
+    "Audio/Source",
+};
+
+static VolumeLevel volume_icon_for(double volume)
```
i’ve actually been writing a display brightness control widget that ends up having roughly the same thing again. Should i preemptively make it even more generic to account for it (or other stuff that might have be similar), and if it’s done soon enough, should i add it to this pr or make another one ?

ammen99 replied:
Better leave it for later, lets finish this PR first and then worry about the next one.

ammen99 commented on src/panel/widgets/wireplumber.cpp:
```diff
@@ -0,0 +1,983 @@
+    update_gestures();
+}
+
+std::unique_ptr<WfWpControl> WfWpControl::copy() // for the face handling
```
I see.

AKArien commented:
@ammen99 i think i have addressed everything pointed out here, but i have a few little ui bugs to fix.

There’s also an issue that i really don’t understand and would like you to keep an eye on in your testing / review : sometimes, under circumstances i cannot nail down, the tray or the menu widgets will segfault, with gdb showing a backtrace in the Watcher destructor for the tray and the toggle_menu method for menu. This never happens without wireplumber in my config, and removing it is always an immediate fix. My widgets are usually like this :
```
widgets_center = separator1 menu separator1
widgets_left = tray wireplumber
widgets_right = notifications launchers battery clock
```
I hardly see how this could be related to my widget, but it seems i’m missing something ? Since removing wireplumber always stops it from happening. Again, i’m sorry if this has very little detail, but it is litterally all that i could find to be consistent.

ammen99 commented:
@AKArien I would strongly recommend compiling wf-shell with address sanitizer in this case. It is likely that you see the result of a memory corruption somewhere and address sanitizer very often catches issues like these.

ammen99 commented on wf-shell.ini.example:
```diff
@@ -130,6 +130,37 @@ menu_logout_command = wayland-logout
 # after scrolling on the volume icon
 volume_display_timeout = 2.5
 
+# Configuration for the wireplumber mixer widget
```
I would add a sentence (or a short paragraph) explaining what this widget does and how it compares to the volume widget. Gotta admit I had no idea what wireplumber is and what this widget was supposed to do before actually reviewing the code.

AKArien replied:
Also changed the name to be more descriptive. If somebody is familiar with audio servers, they’ll probably get what « wp » is , if not, they’ll understand « mixer ».

AKArien commented:
@ammen99 This time should be final. Hopefully.
Updated the initial description again, but basically : fixed long names strecthing the widget, added tooltips, some margins and spacing, an option to put the icons on the left, and changed naming.
Also updated the icons, which applies back to the volume widget, OOR volume is now the emblem-unreadable icon.

Finally, i was wondering if you would like this widget to replace volume in the default configuration ?
With the adoption of pipewire out of the box on pretty much all major distros, it should work on the vast majority of systems, and i would like to think this widget is more polished than the pulseaudio one.
Against that stand the systems on which it will not work out of the box and that some of the functionalities might be confusing the manual by the user’s side. Assuming it all works flawlessly.

ammen99 commented:
> Finally, i was wondering if you would like this widget to replace volume in the default configuration ?
> With the adoption of pipewire out of the box on pretty much all major distros, it should work on the vast majority of systems, and i would like to think this widget is more polished than the pulseaudio one.
> Against that stand the systems on which it will not work out of the box, how non-battle-tested it is, plus that some of the functionalities might be confusing the manual by the user’s side.

Would be best to first let the people test it for a while :) Otherwise I agree that this widget is in pretty much all ways superior to the old widget.
